src/Book.ts

```diff
@@ -0,0 +1,197 @@
+    memo: string,
+    date = null as Date | null,
+    original_journal = null as string | Types.ObjectId | null
+  ): Entry<U, J> {
+    return Entry.write<U, J>(this, memo, date, original_journal);
+  }
+
+  async balance(
+    query: IParseQuery & IPaginationQuery,
+    options = {} as IOptions
+  ): Promise<{ balance: number; notes: number }> {
+    let skip: PipelineStage.Skip | undefined = undefined;
+
+    if (query.perPage) {
+      skip = { $skip: (query.page ? query.page - 1 : 0) * query.perPage };
+
+      delete query.perPage;
+      delete query.page;
+    }
+    const match: PipelineStage.Match = {
+      $match: parseQuery(query, { name: this.name }),
+    };
+
+    const project: PipelineStage.Project = {
+      $project: {
+        debit: "$debit",
+        credit: "$credit",
+        datetime: "$datetime",
+        timestamp: "$timestamp",
+      },
+    };
+    const group: PipelineStage.Group = {
+      $group: {
+        // https://github.com/Automattic/mongoose/pull/11104
+        _id: null as any,
+        credit: {
+          $sum: "$credit",
+        },
+        debit: {
+          $sum: "$debit",
+        },
+        count: {
+          $sum: 1,
+        },
+      },
+    };
+    let result;
+    if (typeof skip !== "undefined") {
+      const sort: PipelineStage.Sort = {
+        $sort: {
+          datetime: -1,
+          timestamp: -1,
+        },
+      };
+      result = (
+        await transactionModel
+          .aggregate([match, project, sort, skip, group], options)
+          .exec()
+      )[0];
+    } else {
+      result = (
+        await transactionModel
+          .aggregate([match, project, group], options)
+          .exec()
+      )[0];
+    }
+    return !result
+      ? {
+          balance: 0,
+          notes: 0,
+        }
+      : {
+          balance: parseFloat(
+            (result.credit - result.debit).toFixed(this.precision)
+          ),
+          notes: result.count,
+        };
+  }
+
+  async ledger<T = U>(
+    query: IParseQuery & IPaginationQuery,
+    populate?: string[] | null,
+    options?: IOptions & { lean?: true }
+  ): Promise<{ results: T[]; total: number }>;
+
+  async ledger<T = U>(
+    query: IParseQuery & IPaginationQuery,
+    populate?: string[] | null,
+    options?: IOptions & { lean?: false }
+  ): Promise<{ results: (Document & T)[]; total: number }>;
+
+  async ledger<T = U>(
+    query: IParseQuery & IPaginationQuery,
+    populate = null as string[] | null,
+    options = {} as IOptions & { lean?: boolean }
+  ): Promise<{ results: T[]; total: number }> {
+    let skip;
+    let limit = 0;
+
+    const { lean = true } = options;
+
+    // Pagination
+    if (query.perPage) {
+      skip = (query.page ? query.page - 1 : 0) * query.perPage;
+      limit = query.perPage;
+    }
+    const filterQuery = parseQuery(query, { name: this.name });
+    const q = transactionModel
+      .find(filterQuery, undefined, options)
+      .lean(lean)
+      .sort({
```

Comment:
We always sort by datetime and timestamp. ParseQuery always breaks our accounts to account_path Array. This makes the index
  schema.index({
    accounts: 1,
    book: 1,
    approved: 1,
    datetime: -1,
    timestamp: -1,
  });

basically useless. 

Also we basically never enforce approved, despite the fact that boolean fields should be the last field in the index. I guess this index is never really used.

I think all the account_path indizes should have datetime and timestamp and the approved flag should be removed or be the last element of the compound index. This could significantly improve the performance of balance and ledger.

Reply:
Added maxAmountPath and modified parseQuery. Now the index can be used.

Reply:
Wow. It is 6x faster to use that index.